File: lib/easy/sum_of_left_leaves.dart
import 'package:dark_coding_challenges/helpers/data_structures/tree_node.dart';

int sumOfLeftLeaves(TreeNode? root) {
  int count = 0;
  List<(TreeNode?, bool)> stack = [(root, false)];
  while (stack.isNotEmpty) {
    final (node, isLeft) = stack.removeLast();
    if (node != null && node.left == null && node.right == null && isLeft) {
      count += node.val ?? 0;
    }
    final right = node?.right;
    final left = node?.left;
    if (right != null) stack.add((right, false));
    if (left != null) stack.add((left, true));
  }
  return count;
}

File: src/main/java/easy/SumOfLeftLeaves.java
package easy;

import helpers.TreeNode;
import java.util.ArrayList;
import java.util.List;

record Pair<T, U>(T a, U b) {}

public class SumOfLeftLeaves {
  public int solve(TreeNode<Integer> root) {
    int count = 0;
    List<Pair<TreeNode<Integer>, Boolean>> stack = new ArrayList<>();
    stack.add(new Pair<>(root, false));
    while (!stack.isEmpty()) {
      Pair<TreeNode<Integer>, Boolean> tuple = stack.removeLast();
      TreeNode<Integer> node = tuple.a();
      Boolean isLeft = tuple.b();
      if (node.left == null && node.right == null && isLeft) {
        count += node.val;
      }
      if (node.left != null) stack.add(new Pair<>(node.left, true));
      if (node.right != null) stack.add(new Pair<>(node.right, false));
    }
    return count;
  }
}
